src/util/pelz_io.c

```diff
@@ -374,17 +391,107 @@ int read_pipe(char *msg)
       {
       case 'c':
         len = strcspn(msg, "\n");
-        path = new_charbuf(len - 11); //the number 11 is used because it the number of chars in "pelz -l -c "
-        memcpy(path.chars, &msg[11], (path.len));
-        free_charbuf(&path);
-        pelz_log(LOG_INFO, "Load cert call not added");
+        path = (char *) malloc((len - 10) * sizeof(char));  //the number 11 is used because it the number of chars in "pelz -l -c "
+        memcpy(path, &msg[11], len);
+
+        if (read_bytes_from_file(path, &data, &data_length))
+        {
+          pelz_log(LOG_ERR, "Unable to read file %s ... exiting", path);
+          free(path);
+          return 1;
+        }
+
+        pelz_log(LOG_DEBUG, "Read bytes from file %s", path);
+        memcpy(path_ext, &path[len - 15], 4); //4 is the set length of .nkl and .ski and 15 is the 11 deducted along with the 4 from the ext
+        pelz_log(LOG_DEBUG, "Path_ext: %s", path_ext);
+        free(path);
+
+        if (memcmp(path_ext, ".ski", 4) == 0)
+        {
+          if (tpm2_kmyth_unseal(data, data_length, &nkl_data, &nkl_data_len, (uint8_t *) authString, auth_string_len,
+              (uint8_t *) ownerAuthPasswd, oa_passwd_len))
+          {
+            pelz_log(LOG_ERR, "TPM unseal failed");
+            free(data);
+            return 1;
+          }
+
+          free(data);
+          if (kmyth_sgx_unseal_nkl(eid, nkl_data, nkl_data_len, &handle))
+          {
+            pelz_log(LOG_ERR, "Unable to unseal contents ... exiting");
+            free(nkl_data);
+            return 1;
+          }
+
+          free(nkl_data);
+        }
+        else if (memcmp(path_ext, ".nkl", 4) == 0)
+        {
+          if (kmyth_sgx_unseal_nkl(eid, data, data_length, &handle))
+          {
+            pelz_log(LOG_ERR, "Unable to unseal contents ... exiting");
+            free(data);
+            free(path);
+            return 1;
+          }
+
+          free(data);
+        }
+
+        pelz_log(LOG_INFO, "Load cert call not finished");
         return 0;
       case 'p':
         len = strcspn(msg, "\n");
-        path = new_charbuf(len - 11); //the number 11 is used because it the number of chars in "pelz -l -p "
-        memcpy(path.chars, &msg[11], (path.len));
-        free_charbuf(&path);
-        pelz_log(LOG_INFO, "Load private call not added");
+        path = (char *) malloc((len - 10) * sizeof(char));  //the number 10 is used because it the number of chars in "pelz -l -p " - 1
+        memcpy(path, &msg[11], len);
+
+        if (read_bytes_from_file(path, &data, &data_length))
+        {
+          pelz_log(LOG_ERR, "Unable to read file %s ... exiting", path);
+          free(path);
+          return 1;
+        }
+
+        pelz_log(LOG_DEBUG, "Read bytes from file %s", path);
+        memcpy(path_ext, &path[len - 15], 4); //4 is the set length of .nkl and .ski and 15 is the 11 deducted along with the 4 from the ext
```

Comment:
Same thing here, cleaner to use strtok_r and set a variable rather than always carrying the extension around.

Reply:
Ignore this, it's just wrong.